{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "YWc5M66hJ1Gi"
   },
   "source": [
    "# Lab report\n",
    "\n",
    "please leave all comments in the notebook cells"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "JqvjvBLH0g95"
   },
   "source": [
    "## 16.10.22 Day first\n",
    "\n",
    "### Step 1\n",
    "\n",
    "At first, I counted a number of lines in .fastq files.\n",
    "\n",
    "```bash \n",
    "zcat <path_to_file>/amp_res_1.fastq.gz | wc -l\n",
    "> 1823504\n",
    "zcat <path_to_file>/amp_res_2.fastq.gz | wc -l\n",
    "> 1823504\n",
    "```\n",
    "`<path_to_file>` means absolute (or sometimes relative path to file)\n",
    "\n",
    "It means that the number of lines is equal in both files. All is ok. If we need to know how many reads are in files, we need to divide this number by four (because there are four lines to one read in .fastq). I tried to write the previous command in the python variable. It did something unfriendly: casts number to string and covered them in a list.\n",
    "\n",
    "```python\n",
    "a = !zcat <path_to_file>/amp_res_1.fastq.gz | wc -l\n",
    "print(a)\n",
    "# ['1823504']\n",
    "```\n",
    "After this, I tried to cast the variable into \"int\" format. It was ok. Then I counted the number of reads.\n",
    "\n",
    "```python\n",
    "a = int(a[0])\n",
    "print(a)\n",
    "# 1823504\n",
    "\n",
    "```\n",
    "```python\n",
    "lines_fastq1 = !zcat <path_to_file>/amp_res_1.fastq.gz | wc -l\n",
    "lines_fastq2 = !zcat <path_to_file>/amp_res_2.fastq.gz | wc -l\n",
    "reads_fastq1 = int(int(lines_fastq1[0])/4)\n",
    "reads_fastq2 = int(int(lines_fastq2[0])/4)\n",
    "print(reads_fastq1, reads_fastq2)\n",
    "# 455876 455876\n",
    "\n",
    "```\n",
    "\n",
    "The number of reads is equal. It means, now we haven't problems with the discordance of two files."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "3o2enoJv3LB1"
   },
   "source": [
    "### Step 2\n",
    "\n",
    "Then I use fastqc (v. 0.11.9) to take some information about reads. I save fastqc's output in the project's directory.\n",
    "\n",
    "```bash\n",
    "!fastqc -o . <path_to_file>/amp_res_1.fastq.gz  <path_to_file>/amp_res_2.fastq.gz\n",
    "```\n",
    "\n",
    "_fastqc checks some parameters of reads: quality, GC-content, level of duplications, adaptor content and many others_\n",
    "\n",
    "Our basic statistic matches what we calculated for the number of reads.\n",
    "\n",
    "We have some warnings from fastq, related to the per base sequence quality and per tile sequence quality in the first (forward reads) file and per base sequence quality in the second (backward reads).\n",
    "\n",
    "The loss of quality at the finish is a common situation associated with sequencing quality loss of Illumina at the end of a read. It's usually not a problem.\n",
    "\n",
    "Warnings from per tile sequence quality on forward reads are more problematical. Patterns of quality loss are associated with air bubbles at the up-left part of the figure. Narrow lines in the right part of the figure may be associated with dirt on the tile."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "WOUcCSEggkDg"
   },
   "source": [
    "### Step 3\n",
    "\n",
    "After this trimmomatic (v. 0.39) was installed.\n",
    "\n",
    "At first, I needed to know in what directory the tool is situated (it is important for using the tool). I used recommended command:\n",
    "\n",
    "```bash\n",
    "dpkg -L trimmomatic | grep '.jar'\n",
    "#/usr/share/java/trimmomatic-0.39.jar\n",
    "#/usr/share/java/trimmomatic.jar\n",
    "```\n",
    "\n",
    "Next script trimmed reads with a quality loss on the trailing end. We excluded reads with quality below 20 at the start and the end, with quality mean below 20 in the 10-line wide window on tile and with the length of reads lesser than 20 bases.\n",
    "\n",
    "```bash\n",
    "ava -jar /usr/share/java/trimmomatic-0.39.jar PE -phred33 <path_to_file>/amp_res_1.fastq.gz <path_to_file>/amp_res_2.fastq.gz <path_to_file>/po1.fq <path_to_file>/upo1.fq <path_to_file>/po2.fq <path_to_file>/upo2.fq LEADING:20 TRAILING:20 SLIDINGWINDOW:10:20 MINLEN:20\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "jB06ly2v0_do"
   },
   "source": [
    "## 18.10.22 Day second\n",
    "\n",
    "### Step 4\n",
    "\n",
    "I used bwa (v. 0.7.17-r1188) for indexing the reference fasta sequence.\n",
    "\n",
    "```bash\n",
    "bwa index <path_to_file>/GCF_000005845.2_ASM584v2_genomic.fna.gz\n",
    "```\n",
    "_bwa is an utility for making Burrowse-Wheeler transformation for the next mapping of big amounts of sequences._\n",
    "\n",
    "Output after ```ls``` command was\n",
    "\n",
    "```bash\n",
    "amp_res_1.fastq.gz\n",
    "amp_res_2.fastq.gz\n",
    "GCF_000005845.2_ASM584v2_genomic.fna.gz\n",
    "GCF_000005845.2_ASM584v2_genomic.fna.gz.amb   #\n",
    "GCF_000005845.2_ASM584v2_genomic.fna.gz.ann   #\n",
    "GCF_000005845.2_ASM584v2_genomic.fna.gz.bwt   #\n",
    "GCF_000005845.2_ASM584v2_genomic.fna.gz.pac   #\n",
    "GCF_000005845.2_ASM584v2_genomic.fna.gz.sa    #\n",
    "GCF_000005845.2_ASM584v2_genomic.gff.gz\n",
    "\n",
    "```\n",
    "\n",
    "For reading alignments, I used bwa mem.\n",
    "```bash\n",
    "bwa mem path_to_file/GCF_000005845.2_ASM584v2_genomic.fna.gz <path_to_file>po1.fq <path_to_file>po2.fq > alignments.sam\n",
    "```\n",
    "\n",
    "```\n",
    "@SQ\tSN:lcl|NC_000913.3_cds_NP_414542.1_1\tLN:66\n",
    "@SQ\tSN:lcl|NC_000913.3_cds_NP_414543.1_2\tLN:2463\n",
    "@SQ\tSN:lcl|NC_000913.3_cds_NP_414544.1_3\tLN:933\n",
    "@SQ\tSN:lcl|NC_000913.3_cds_NP_414545.1_4\tLN:1287\n",
    "@SQ\tSN:lcl|NC_000913.3_cds_NP_414546.1_5\tLN:297\n",
    "@SQ\tSN:lcl|NC_000913.3_cds_NP_414547.1_6\tLN:777\n",
    "@SQ\tSN:lcl|NC_000913.3_cds_NP_414548.1_7\tLN:1431\n",
    "@SQ\tSN:lcl|NC_000913.3_cds_NP_414549.1_8\tLN:954\n",
    "@SQ\tSN:lcl|NC_000913.3_cds_NP_414550.1_9\tLN:588\n",
    "@SQ\tSN:lcl|NC_000913.3_cds_NP_414551.1_10\tLN:567\n",
    "```\n",
    "\n",
    "For zipping this file I used samtools (v. 1.16.1).\n",
    "\n",
    "```bash\n",
    "samtools view -S -b alignments.sam > alignments.bam\n",
    "```\n",
    "\n",
    "For indexing the BAM files I used samtools too...\n",
    "\n",
    "```bash\n",
    "samtools sort alignments.bam -o al_sorted.bam\n",
    "```\n",
    "\n",
    "... and then I indexed this file for faster working with it.\n",
    "\n",
    "```bash\n",
    "samtools index al_sorted.bam\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "ImmQYHp3QT-x"
   },
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "EaLvIsbfPRBL"
   },
   "source": [
    "## 20.10.22 Day third\n",
    "### Step 5\n",
    "\n",
    "All results was succesfully reproduced by Kirill Sizov. Head of alignments.sam was added to lab report of the previous day. He give some recomendations for previous lab  notes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "2YP12-wQPG2N"
   },
   "source": [
    "## 21.10.22 Day fourth\n",
    "\n",
    "# Step 6\n",
    "\n",
    "For distinguishing actual mutations from the sequencing errors I used mpileup tool.\n",
    "\n",
    "```bash\n",
    "samtools mpileup -f ./raw_data/GCF_000005845.2_ASM584v2_genomic.fna al_sorted.bam >  my.mpileup\n",
    "```\n",
    "\n",
    "First ten rows of data in my.mpileup:\n",
    "\n",
    "```bash\n",
    "cat my.mpileup | head -10\n",
    "#NC_000913.3\t1\tA\t5\t^].^].^],^],^],\t>;EEB\n",
    "#NC_000913.3\t2\tG\t7\t...,,,,\tE1=JII0\n",
    "#NC_000913.3\t3\tC\t7\t...,,,,\tH9=IGH8\n",
    "#NC_000913.3\t4\tT\t7\t...,,,,\tE;CGCD9\n",
    "#NC_000913.3\t5\tT\t7\t...,,,,\tC<EIJJ:\n",
    "#NC_000913.3\t6\tT\t7\t...,,,,\tE=EIII1\n",
    "#NC_000913.3\t7\tT\t7\t...,,,,\tC@EHHG1\n",
    "#NC_000913.3\t8\tC\t6\t...,,,\tDE>JDJ\n",
    "#NC_000913.3\t9\tA\t6\t...,,,\tDH?JHI\n",
    "#NC_000913.3\t10\tT\t7\t...,,,,\tFGCIHEH\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "-IBsnILXSfX9"
   },
   "source": [
    "## 23.10.22 Day fifth.\n",
    "### Step 7\n",
    "\n",
    "I used VarScan (v. 2.3.9) for differiantiating sequencing errors from SNPs.\n",
    "\n",
    "```bash\n",
    "java -jar ../VarScan.v2.3.9.jar  mpileup2snp my.mpileup --min-var-freq 0.1 --variants --output-vcf 1 > VarScan_results.vcf\n",
    "```\n",
    "\n",
    "After this, I repeated scanning with `--min-var-freq 0.05 ` and `--min-var-freq 0.5 `, but had the same result in IGV."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "CzaiUw7qUSb_"
   },
   "source": [
    "### Step 8\n",
    "\n",
    "For identification locuses of mutations I used SnpEff (v. 5.1d).\n",
    "\n",
    "Firstly, I create database for E.coli K12. Annotation and sequence were downloaded from [GeneBank](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.gbff.gz)\n",
    "\n",
    "Then I created in working directory new file `vim snpeff.config` with \n",
    "> k12.genome : ecoli_K12\n",
    "\n",
    "content.\n",
    "\n",
    "File was moved to other directory, unpacked and renamed to genes.gbk\n",
    "\n",
    "I created database by command `snpEff build -genbank -v k12`, when I was in \"data\" directory. I had a problems with `snpEff ann k12 VarScan_results.vcf` command from guideline, then I try to solve the problem manually.\n",
    "\n",
    "Upd.: my coautor does this step, then we have file for checking ourselves. All was correct.\n",
    "Upd.2.0: the solution of problem was decribed in workshop's chat.\n",
    "\n",
    "It was very important to check the strand orientation in gene of interest and tap \"flip strand\" to change aminoacid sequence of the annotation.\n",
    "\n",
    "We have SNP (G instead C) in the middle position of A. It changes alanin (GCC) to glycin (GGC). We have missens mutation in CDS. Full information about this locus is below.\n",
    "``` \n",
    "NC_000913.3:91413-93179\n",
    "\n",
    "Type: gene\n",
    "ID: gene-b0084\n",
    "Dbxref: ASAP:ABE-0000309,ECOCYC:EG10341,GeneID:944799\n",
    "Name: ftsI\n",
    "gbkey: Gene\n",
    "gene: ftsI\n",
    "gene_biotype: protein_coding\n",
    "gene_synonym: ECK0085,pbpB,sep\n",
    "locus_tag: b0084\n",
    "---------------------------\n",
    "Type: CDS\n",
    "ID: cds-NP_414626.1\n",
    "Parent: gene-b0084\n",
    "Dbxref: UniProtKB/Swiss-Prot:P0AD68,Genbank:NP_414626.1,AS\n",
    "AP:ABE-0000309,ECOCYC:EG10341,GeneID:944799\n",
    "Name: NP_414626.1\n",
    "gbkey: CDS\n",
    "gene: ftsI\n",
    "locus_tag: b0084\n",
    "orig_transcript_id: gnl|b0084|mrna.NP_414626\n",
    "product: peptidoglycan DD-transpeptidase FtsI\n",
    "protein_id: NP_414626.1\n",
    "transl_table: 11\n",
    "```\n",
    "Second locus was very strange. After flipping strands it shows that A substituted... A. I screened IGV for futher consultation and continue to watching.\n",
    "\n",
    "Third locus didn't contain protein-encoding gene, but had small RNA exon and substitution G instead T. I have some ambiguous ideas about this mutation and continue to watching IGV results.\n",
    "```\n",
    "NC_000913.3:852725-853064\n",
    "\n",
    "Type: gene\n",
    "ID: gene-b4416\n",
    "Dbxref: ASAP:ABE-0047240,ECOCYC:G0-8881,GeneID:2847681\n",
    "Name: rybA\n",
    "gbkey: Gene\n",
    "gene: rybA\n",
    "gene_biotype: ncRNA\n",
    "gene_synonym: ECK0806\n",
    "locus_tag: b4416\n",
    "```\n",
    "```\n",
    "NC_000913.3:852725-853064\n",
    "\n",
    "Type: ncRNA\n",
    "ID: rna-b4416\n",
    "Parent: gene-b4416\n",
    "Dbxref: ASAP:ABE-0047240,ECOCYC:G0-8881,GeneID:2847681\n",
    "gbkey: ncRNA\n",
    "gene: rybA\n",
    "locus_tag: b4416\n",
    "product: small RNA RybA\n",
    "---------------------------\n",
    "Type: exon\n",
    "ID: exon-b4416-1\n",
    "Parent: rna-b4416\n",
    "Dbxref: ASAP:ABE-0047240,ECOCYC:G0-8881,GeneID:2847681\n",
    "gbkey: ncRNA\n",
    "gene: rybA\n",
    "locus_tag: b4416\n",
    "product: small RNA RybA\n",
    "\n",
    "```\n",
    "\n",
    "Fourth locus contained substituition A to G. We have missens mutation because replacing GGT to GAT changes glycin to aspartic acid. Full information about this locus is below.\n",
    "\n",
    "```\n",
    "NC_000913.3:1905688-1906254\n",
    "\n",
    "Type: gene\n",
    "ID: gene-b1821\n",
    "Dbxref: ASAP:ABE-0006065,ECOCYC:G6999,GeneID:946341\n",
    "Name: mntP\n",
    "gbkey: Gene\n",
    "gene: mntP\n",
    "gene_biotype: protein_coding\n",
    "gene_synonym: ECK1819,yebN\n",
    "locus_tag: b1821\n",
    "---------------------------\n",
    "Type: CDS\n",
    "ID: cds-NP_416335.4\n",
    "Parent: gene-b1821\n",
    "Dbxref: UniProtKB/Swiss-Prot:P76264,Genbank:NP_416335.4,AS\n",
    "AP:ABE-0006065,ECOCYC:G6999,GeneID:946341\n",
    "Name: NP_416335.4\n",
    "gbkey: CDS\n",
    "gene: mntP\n",
    "locus_tag: b1821\n",
    "orig_transcript_id: gnl|b1821|mrna.NP_416335\n",
    "product: Mn(2( )) exporter\n",
    "protein_id: NP_416335.4\n",
    "transl_table: 11\n",
    "\n",
    "```\n",
    "\n",
    "Fifth is substitution C to T. We have missens-mutation, because exchange GTA to GCA substitutes valine to alanine.\n",
    "```\n",
    "NC_000913.3:3534516-3535868\n",
    "\n",
    "Type: gene\n",
    "ID: gene-b3404\n",
    "Dbxref: ASAP:ABE-0011108,ECOCYC:EG10269,GeneID:947272\n",
    "Name: envZ\n",
    "gbkey: Gene\n",
    "gene: envZ\n",
    "gene_biotype: protein_coding\n",
    "gene_synonym: ECK3391,ompB,perA,tpo\n",
    "locus_tag: b3404\n",
    "---------------------------\n",
    "Type: CDS\n",
    "ID: cds-NP_417863.1\n",
    "Parent: gene-b3404\n",
    "Dbxref: UniProtKB/Swiss-Prot:P0AEJ4,Genbank:NP_417863.1,AS\n",
    "AP:ABE-0011108,ECOCYC:EG10269,GeneID:947272\n",
    "Name: NP_417863.1\n",
    "gbkey: CDS\n",
    "gene: envZ\n",
    "locus_tag: b3404\n",
    "orig_transcript_id: gnl|b3404|mrna.NP_417863\n",
    "product: sensor histidine kinase EnvZ\n",
    "protein_id: NP_417863.1\n",
    "transl_table: 11\n",
    "\n",
    "```\n",
    "\n",
    "\n",
    "Sixth is substitution T to C. We have samesense-mutation, because GCC and GCT code the same aminoacid - alanine."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "_mWQf6L2Oi22"
   },
   "source": [
    "## 25-26.10.22 Days sixth and seventh\n",
    "### Step nine\n",
    "\n",
    "Writing an mini-article. We used over-leaf. Dixi et animam levavi.\n"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
