{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# Project 2\n",
        "Lab journal\n",
        "\n",
        "Bogdan Sotnikov"
      ],
      "metadata": {
        "id": "Vkbs3wm_-_iJ"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Day 1 20.11.22\n",
        "\n",
        "### Step 1. The preparing to the investigation.\n",
        "\n",
        "We created virtual environment for this project.\n",
        "\n",
        "```bash\n",
        "conda create -n BI_proj3\n",
        "```\n",
        "\n",
        "We downloaded three complects of paired reads (by one forward and reverse  in each complect).\n",
        "\n",
        "We used bash command for downloading this data\n",
        "\n",
        "```bash\n",
        "wget https://d28rh4a8wq0iu5.cloudfront.net/bioinfo/SRR292678sub_S1_L001_R1_001.fastq.gz \n",
        "\n",
        "wget https://d28rh4a8wq0iu5.cloudfront.net/bioinfo/SRR292678sub_S1_L001_R2_001.fastq.gz \n",
        "\n",
        "wget https://d28rh4a8wq0iu5.cloudfront.net/bioinfo/SRR292862_S2_L001_R1_001.fastq.gz \n",
        "\n",
        "wget https://d28rh4a8wq0iu5.cloudfront.net/bioinfo/SRR292862_S2_L001_R2_001.fastq.gz \n",
        "\n",
        "wget https://d28rh4a8wq0iu5.cloudfront.net/bioinfo/SRR292770_S1_L001_R1_001.fastq.gz \n",
        "\n",
        "wget https://d28rh4a8wq0iu5.cloudfront.net/bioinfo/SRR292770_S1_L001_R2_001.fastq.gz \n",
        "\n",
        "```\n",
        "The first pair comtained paired-end reads and the second two pairs contained mate-paired reads."
      ],
      "metadata": {
        "id": "tcpfWeN8_g8Y"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Step 2. The data analysis.\n",
        "\n",
        "We used fastqc (v0.11.9) to check our data's quality. Quality was good, without any abnormalities.\n",
        "\n",
        "```bash\n",
        "fastqc *_001.fastq.gz \n",
        "```\n",
        "\n",
        "For k-mer analyzing we used jellyfish (v. 2.3.0).\n",
        "We used it with recomended options: m = 31 (number of k-mers), -C for ignoring strands and 10000000 for nu,ber of entries.\n",
        "\n",
        "```bash\n",
        "jellyfish count -m 31 -o jfoutput_1 -s 10000000 -C SRR292678sub_S1_L001_R1_001.fastq SRR292678sub_S1_L001_R2_001.fastq\n",
        "```\n",
        "Then we found more detailed explanation for working with jelyyfish and created another file.\n",
        "\n",
        "```bash\n",
        "jellyfish count -m 31 -o jfoutput_31 -s 50000000 -C --min-quality=20 --quality-start=33 SRR292678sub*\n",
        "```\n",
        "\n",
        "We added options obout quality-coding (PHRED-33) and about threshold for quality.\n",
        "\n",
        "Then we used jellyfish histo for creating human-readable data about k-mers.\n",
        "\n",
        "On the next stage we visualised k-mer distribution in R (v. 4.2.1).\n",
        "\n",
        "```r\n",
        "setwd(\"<path_to_directory>\")\n",
        "library(ggplot2)\n",
        "spec1_31 <- read.table(\"jfoutput_31.histo\")\n",
        "plot(spec1_31[2:200,],type=\"l\")\n",
        "\n",
        "ggplot(spec1_31[2:200,])+\n",
        "  geom_point(aes(x=spec1_31[2:200,]$V1, y=spec1_31[2:200,]$V2))+\n",
        "  xlab(label = \"Frequency of k-mer\")+\n",
        "  ylab(label = \"Frequency of k-mer's frequency\")\n",
        "\n",
        "total_kmer <- sum(as.numeric(spec1_31[3:nrow(spec1_31), 1]*spec1_31[3:nrow(spec1_31), 2]))\n",
        "total_kmer\n",
        "peak_kmer <- total_kmer/48\n",
        "peak_kmer\n",
        "\n",
        "\n",
        "# 269318297\n",
        "# 5610798\n",
        "```\n",
        "\n",
        "!!!Вставить картинку!!!\n",
        "\n",
        "We found inflection point (3), total number of k-mers (269318297) and peak position (48). We can compute genome size as  T/(M*L)/(L-K+1), where M: Kmer peak, K: Kmer-size, L: avg read length T: Total bases. Divider is coverage depth.\n",
        "Genome size is equal (5499346*90)/(48*90)/(90-31+1) = 6874183 = 6.9 G.\n",
        "Another genome size equation is total number of k-mers/peak position = 5.6 G. Second estimation looks more realistically.\n",
        "\n",
        " and genome size (5610798 bases = 5.6 Gb)."
      ],
      "metadata": {
        "id": "D8QzgTIoDtET"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Step 3. Genome assembling.\n",
        "\n",
        "#### Substep 3.1. 1-library assembling.\n",
        "\n",
        "For assembling E. coli X genome from paired reads we used SPAdes (v3.15.5).\n",
        "\n",
        "I tried to start spades with below-writed options:\n",
        "\n",
        "```bash\n",
        "spades.py --isolate --pe1-1 SRR292678sub_S1_L001_R1_001.fastq --pe2-1 SRR292678sub_S1_L001_R2_001.fastq -o <path_to_directory>\n",
        "```\n",
        "\n",
        "Spades didn't work with these options. Some time later I found one mistake: I wrote \"--pe2-1\" in place of \"--pe1-2\". When I corrected it, process had started.\n",
        "\n",
        "The correct command was \n",
        "\n",
        "```bash\n",
        "spades.py --isolate --pe1-1 SRR292678sub_S1_L001_R1_001.fastq --pe1-2 SRR292678sub_S1_L001_R2_001.fastq -o <path_to_directory>\n",
        "```\n",
        "\n",
        "But it wasn't finally correct too, beacuse I printed path from \"/\" and spades created new path. I had needed to move files to working directory (and I did it).\n",
        "\n",
        "For assesing quality of assembling we used QUAST (v. 2.31). Quast was uncompartible with python 3.11, thats why we created new conda environment.\n",
        "\n",
        "```bash\n",
        "conda create -n BI_proj39 python=3.9\n",
        "```\n",
        "We started this tool with below-described options.\n",
        "\n",
        "```bash\n",
        "quast.py contigs.fasta -o ~/bioinf/ngs/proj3\n",
        "```\n",
        "\n",
        "Information from output was \n",
        "```\n",
        "All statistics are based on contigs of size >= 500 bp, unless otherwise noted (e.g., \"# contigs (>= 0 bp)\" and \"Total length (>= 0 bp)\" include all contigs).\n",
        "\n",
        "Assembly                    contigs \n",
        "# contigs (>= 0 bp)         885     \n",
        "# contigs (>= 1000 bp)      127     \n",
        "# contigs (>= 5000 bp)      86      \n",
        "# contigs (>= 10000 bp)     73      \n",
        "# contigs (>= 25000 bp)     54      \n",
        "# contigs (>= 50000 bp)     32      \n",
        "Total length (>= 0 bp)      5363438 \n",
        "Total length (>= 1000 bp)   5227528 \n",
        "Total length (>= 5000 bp)   5112460 \n",
        "Total length (>= 10000 bp)  5024868 \n",
        "Total length (>= 25000 bp)  4704658 \n",
        "Total length (>= 50000 bp)  3898474 \n",
        "# contigs                   162     \n",
        "Largest contig              300784  \n",
        "Total length                5252199 \n",
        "GC (%)                      50.54   \n",
        "N50                         114227  \n",
        "N90                         24915   \n",
        "auN                         126907.1\n",
        "L50                         15      \n",
        "L90                         55      \n",
        "# N's per 100 kbp           0.00  \n",
        "```\n",
        "\n",
        "It was slight different data for scaffolds:\n",
        "\n",
        "```\n",
        "All statistics are based on contigs of size >= 500 bp, unless otherwise noted (e.g., \"# contigs (>= 0 bp)\" and \"Total length (>= 0 bp)\" include all contigs).\n",
        "\n",
        "Assembly                    scaffolds\n",
        "# contigs (>= 0 bp)         868      \n",
        "# contigs (>= 1000 bp)      123      \n",
        "# contigs (>= 5000 bp)      83       \n",
        "# contigs (>= 10000 bp)     70       \n",
        "# contigs (>= 25000 bp)     53       \n",
        "# contigs (>= 50000 bp)     32       \n",
        "Total length (>= 0 bp)      5365096  \n",
        "Total length (>= 1000 bp)   5228807  \n",
        "Total length (>= 5000 bp)   5115861  \n",
        "Total length (>= 10000 bp)  5026097  \n",
        "Total length (>= 25000 bp)  4751300  \n",
        "Total length (>= 50000 bp)  3989113  \n",
        "# contigs                   168      \n",
        "Largest contig              300784   \n",
        "Total length                5260240  \n",
        "GC (%)                      50.52    \n",
        "N50                         114589   \n",
        "N90                         25454    \n",
        "auN                         130855.3 \n",
        "L50                         14       \n",
        "L90                         53       \n",
        "# N's per 100 kbp           32.47 \n",
        "```\n",
        "\n",
        "#### Substep 3.2. 3-library assembling.\n",
        "\n",
        "For assembling with matepairs we used below-writed code:\n",
        "\n",
        "```bash\n",
        "spades.py --isolate --pe1-1 SRR292678sub_S1_L001_R1_001.fastq.gz --pe1-2 SRR292678sub_S1_L001_R2_001.fastq.gz --mp1-1 SRR292770_S1_L001_R2_001.fastq.gz --mp1-2 SRR292770_S1_L001_R1_001.fastq.gz --mp2-1 SRR292862_S2_L001_R2_001.fastq.gz --mp2-2 SRR292862_S2_L001_R1_001.fastq.gz -o <path_for_output>\n",
        "```\n",
        "\n",
        "We assessed quality of assembling by quast:\n",
        "\n",
        "```bash\n",
        "quast.py contigs.fasta -o <path_for_output>\n",
        "```\n",
        "\n",
        "```bash\n",
        "quast.py scaffolds.fasta -o <path_for_output>\n",
        "```\n",
        "\n",
        "with two outputs respectively:\n",
        "\n",
        "```\n",
        "Assembly                    contigs \n",
        "# contigs (>= 0 bp)         1154    \n",
        "# contigs (>= 1000 bp)      126     \n",
        "# contigs (>= 5000 bp)      66      \n",
        "# contigs (>= 10000 bp)     54      \n",
        "# contigs (>= 25000 bp)     44      \n",
        "# contigs (>= 50000 bp)     28      \n",
        "Total length (>= 0 bp)      5402541 \n",
        "Total length (>= 1000 bp)   5221609 \n",
        "Total length (>= 5000 bp)   5090134 \n",
        "Total length (>= 10000 bp)  5011459 \n",
        "Total length (>= 25000 bp)  4846400 \n",
        "Total length (>= 50000 bp)  4280675 \n",
        "# contigs                   169     \n",
        "Largest contig              391919  \n",
        "Total length                5251556 \n",
        "GC (%)                      50.54   \n",
        "N50                         157905  \n",
        "N90                         29147   \n",
        "auN                         177788.0\n",
        "L50                         11      \n",
        "L90                         40      \n",
        "# N's per 100 kbp           0.00\n",
        "```\n",
        "\n",
        "```\n",
        "Assembly                    scaffolds\n",
        "# contigs (>= 0 bp)         1070     \n",
        "# contigs (>= 1000 bp)      56       \n",
        "# contigs (>= 5000 bp)      13       \n",
        "# contigs (>= 10000 bp)     12       \n",
        "# contigs (>= 25000 bp)     11       \n",
        "# contigs (>= 50000 bp)     11       \n",
        "Total length (>= 0 bp)      5531597  \n",
        "Total length (>= 1000 bp)   5351356  \n",
        "Total length (>= 5000 bp)   5266301  \n",
        "Total length (>= 10000 bp)  5259820  \n",
        "Total length (>= 25000 bp)  5248736  \n",
        "Total length (>= 50000 bp)  5248736  \n",
        "# contigs                   108      \n",
        "Largest contig              1046981  \n",
        "Total length                5386995  \n",
        "GC (%)                      50.52    \n",
        "N50                         741159   \n",
        "N90                         426441   \n",
        "auN                         719107.4 \n",
        "L50                         3        \n",
        "L90                         7        \n",
        "# N's per 100 kbp           2395.70 \n",
        "```"
      ],
      "metadata": {
        "id": "tdFChDEqooDU"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Day 2. 21.11.22\n",
        "\n",
        "Step 4. Annotating.\n",
        "\n",
        "For annotating assembled genome we used prokka (v. 2.31). It were some problems with installation, but, as it looks, now all is coreect. (UPD - it didn't works).\n",
        "\n",
        "I looked log and found that orikka conflicted with QUAST. Then I created new conda environment.\n",
        "\n",
        "```bash\n",
        "conda install -c bioconda prokka\n",
        "```\n",
        "There were no warnings, but prokka didn't works too.\n",
        "\n",
        "Then I tried a next way:\n",
        "\n",
        "```bash\n",
        "sudo apt-get install libdatetime-perl libxml-simple-perl libdigest-md5-perl git default-jre bioperl\n",
        "sudo cpan Bio::Perl\n",
        "git clone https://github.com/tseemann/prokka.git $HOME/prokka\n",
        "$HOME/prokka/bin/prokka --setupdb\n",
        "```\n",
        "New attempt leaded to new failure.\n",
        "\n",
        "We used the next script\n",
        "\n",
        "```bash\n",
        "sudo apt-get -y update\n",
        "sudo apt-get install bioperl libxml-simple-perl default-jre git curl\n",
        "\n",
        "\n",
        "\n",
        "git clone https://github.com/tseemann/prokka.git\n",
        "export PATH=$PWD/prokka/bin:$PATH\n",
        "prokka --setupdb\n",
        "prokka --version\n",
        "```\n",
        "\n",
        "We used prokka FROM WORKING DIRECTORY, WHERE \"PROKKA\" FOLDER WAS SITUATED (sorry for caps) with below-writed script.\n",
        "\n",
        "```bash\n",
        "git clone https://github.com/tseemann/prokka.git\n",
        "export PATH=$PWD/prokka/bin:$PATH\n",
        "prokka --setupdb\n",
        "prokka --version\n",
        "```\n",
        "\n",
        "BUT IT DIDN'T WORK(sorry for caps)\n",
        "\n",
        "I created new one venv.\n",
        "\n",
        "```bash\n",
        "conda create -n prokka_env -c conda-forge -c bioconda prokka\n",
        "```\n",
        "\n",
        "And started prokka (v. 1.14.6)\n",
        "\n",
        "```bash\n",
        "prokka ./spades_3library/scaffolds.fasta\n",
        "```\n"
      ],
      "metadata": {
        "id": "7fC4muPtz6gs"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Step 5. Finding 16S.\n",
        "\n",
        "For finding 16S (we need them to estimate more relative E.coli strains) we used barrnap (v 0.9). We entered this tool with default parameters.\n",
        "\n",
        "```bash\n",
        "barrnap scaffolds.fasta > 16srna\n",
        "```\n",
        "\n",
        "With the output:\n",
        "\n",
        "```\n",
        "##gff-version 3\n",
        "NODE_1_length_1046981_cov_69.027703\tbarrnap:0.9\trRNA\t405650\t405760\t2.3e-11\t+\t.\tName=5S_rRNA;product=5S ribosomal RNA\n",
        "NODE_266_length_223_cov_0.720238\tbarrnap:0.9\trRNA\t19\t129\t5.7e-18\t+\t.\tName=5S_rRNA;product=5S ribosomal RNA\n",
        "NODE_2_length_981131_cov_67.425569\tbarrnap:0.9\trRNA\t526837\t528374\t0-\t.\tName=16S_rRNA;product=16S ribosomal RNA\n",
        "NODE_3_length_741159_cov_70.451946\tbarrnap:0.9\trRNA\t353641\t355178\t0+\t.\tName=16S_rRNA;product=16S ribosomal RNA\n",
        "NODE_56_length_1031_cov_481.502049\tbarrnap:0.9\trRNA\t9\t1026\t5.9e-258\t-\t.\tName=23S_rRNA;product=23S ribosomal RNA (partial);note=aligned only 31 percent of the 23S ribosomal RNA\n",
        "NODE_5_length_546617_cov_64.838179\tbarrnap:0.9\trRNA\t124551\t124661\t2.3e-11\t+\t.\tName=5S_rRNA;product=5S ribosomal RNA\n",
        "NODE_5_length_546617_cov_64.838179\tbarrnap:0.9\trRNA\t221914\t222024\t2.3e-11\t+\t.\tName=5S_rRNA;product=5S ribosomal RNA\n",
        "NODE_5_length_546617_cov_64.838179\tbarrnap:0.9\trRNA\t361249\t361359\t2.3e-11\t+\t.\tName=5S_rRNA;product=5S ribosomal RNA\n",
        "NODE_5_length_546617_cov_64.838179\tbarrnap:0.9\trRNA\t402585\t402695\t2.3e-11\t+\t.\tName=5S_rRNA;product=5S ribosomal RNA\n",
        "NODE_76_length_721_cov_0.948949\tbarrnap:0.9\trRNA\t2\t407\t1e-22\t-.\tName=16S_rRNA;product=16S ribosomal RNA (partial);not\n",
        "```\n",
        "\n",
        "For more detailed output I used another command:\n",
        "\n",
        "```bash\n",
        "barrnap -o rrna.fa < scaffolds.fasta > rrna.gff\n",
        "```\n",
        "\n",
        "Then we ran two 16s intervals in blastn with params \"Entrez Query\":1900/01/01:2011/01/01[PDAT].\n",
        "\n",
        "The two nearest organisms were Escherichia coli 55989 chromosome and Shigella flexneri 2a str. 2457T for first 16S sequence; Escherichia coli O26:H11 str. 11368,  \tEscherichia coli 55989, Escherichia coli HS,  \tShigella dysenteriae strain FBD015, Shigella sonnei Ss046 for second 16s sequence; Wallemia ichthyophaga strain CBS 113033 for third 16S sequence.\n",
        "\n",
        "We thank most probable variant is Escherichia coli 55989. The NCBI ID of this bactery is NCBI:txid585055"
      ],
      "metadata": {
        "id": "tCP9XpDh9scx"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "## Day 3. 25.11.22\n",
        "### Step 6. Comparing two E.Coli sequences.\n",
        "\n",
        "I created new virtual environment.\n",
        "\n",
        "```bash\n",
        "conda create -n mauve\n",
        "```\n",
        "\n",
        "Downloaded mauve and old java version.\n",
        "\n",
        "```bash\n",
        "conda install mauve\n",
        "sudo apt install openjdk-8-jdk \n",
        "```\n",
        "\n",
        "\n",
        "We used Mauve (v. ) for this purpose. I added reference file and anotation made by prokka and searching products by pattern \"toxi\".\n",
        "\n",
        "We add reference file 55989.fasta and annotation by prokka.\n",
        "\n",
        "We found shiga related toxins, which were in E. coli X annotation, not in reference. It may cause hemorragic properties of E.coli X. Data about shiga toxin coding parts is below-listed.\n",
        "\n",
        "5307965 5308234 stxB (код на uniprot Q7BQ98 - можно вставить модельку в статью).\n",
        "\n",
        "5308246 5309245 stxA (Q9FBI2)\n",
        "\n",
        "Another interesting insertions:\n",
        "toxic protein HokE 5441456 5441668 (P77091)\n",
        "\n",
        "3493013 3493390 P76364\n",
        "\n",
        "LtxB 4488967 4491069 P23702 (пришла из какого-то аггрибактера).\n",
        "\n",
        "Toxin CCDB 5155511 5155816 P62554\n",
        "\n",
        "Also we have many transposase proteins in E. coli X, which means there are many transposons in it.\n",
        "\n",
        "I have seen many hypothetical proteins, so I used .gbk annotation from project manual.\n",
        "\n",
        "Findings\n",
        "\n",
        "Enterotoxin with long name: 3016232 3019567\n",
        "\n",
        "Another one shiga toxin 3483605 3483874\n",
        "\n",
        "3483886 3484845\n",
        "\n",
        "VapC 5219438 5219866\n",
        "\n",
        "Doc toxin 5224169 5224444"
      ],
      "metadata": {
        "id": "xvcEnUfX-oaE"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Day 4. 26.11.22\n",
        "### Step 7. Finding antibiotic resistance genes.\n",
        "\n",
        "For finding resistance genes I used [ResFinder website](https://cge.food.dtu.dk/services/ResFinder/) v. 4.1 with default options (threshold for % ID - 90%, minimum length - 60%, type of the data - assembled genome/contigs)\n",
        "\n",
        "It is resistant to ampicillin, sulfamethoxazole, trimethoprim, cephalothin\n",
        "piperacillin, tetracycline, amoxicillin, ethidium bromide, cefepime, doxycycline, benzylkonium chloride, ceftriaxone, ceftazidime, ticarcillin, chlorhexidine, cetylpyridinium chloride, cefotaxime, aztreonam  and streptomycin.\n",
        "\n",
        "By comparison reference sequence has resistance only for doxycycline, tetracycline and minocycline.\n",
        "\n",
        "Results means that besides tetracycline resistance E.coli X has resistance for amny beta-lactams, folate pathway antagonists and some antiseptics (quaternary ammonium compounds).\n",
        "\n",
        "I started to find genes, causes this new resistance mechanisms. I used mauve with same parameters, as in previous step.\n",
        "\n",
        "We found beta-lactamase gene TEM (which was absent in reference) in 5108214-5109083 (P62593), and CTX-M1 in 51111970-5112845 (P28585).\n",
        "\n",
        "Also I found new tetracycline resistance gene, which was absent too in reference -- tetA 5281235-5282434 (P02981).\n",
        "\n",
        "We need some additional information to interpret significanse of presence dhfrI (dihydrofolate reductase type 1) 4957362-5049029 (P00382).\n",
        "\n",
        "\n",
        "I found transposone resolvase and transposase near two bla-genes. It may account that this genes came from transposones."
      ],
      "metadata": {
        "id": "yCH_O7bmp8ss"
      }
    }
  ]
}